src/scheduling/cost_models/simulated_quincy_cost_model.cc

```diff
@@ -0,0 +1,358 @@
+// The Firmament project
+// Copyright (c) 2014 Malte Schwarzkopf <malte.schwarzkopf@cl.cam.ac.uk>
+//
+// Quincy scheduling cost model, as described in the SOSP 2009 paper.
+
+#include "scheduling/cost_models/simulated_quincy_cost_model.h"
+
+#include <set>
+#include <string>
+#include <unordered_map>
+
+#include "base/common.h"
+#include "base/types.h"
+#include "misc/map-util.h"
+#include "misc/utils.h"
+#include "scheduling/common.h"
+#include "scheduling/cost_models/flow_scheduling_cost_model_interface.h"
+#include "scheduling/knowledge_base.h"
+
+namespace firmament {
+
+SimulatedQuincyCostModel::SimulatedQuincyCostModel(
+    shared_ptr<ResourceMap_t> resource_map,
+    shared_ptr<JobMap_t> job_map,
+    shared_ptr<TaskMap_t> task_map,
+    unordered_map<TaskID_t, ResourceID_t> *task_bindings,
+    unordered_set<ResourceID_t,
+      boost::hash<boost::uuids::uuid>>* leaf_res_ids,
+    KnowledgeBase* kb,
+    SimulatedDFS* dfs,
+    GoogleRuntimeDistribution *runtime_distribution,
+    GoogleBlockDistribution *block_distribution,
+		double delta_preferred_machine,
+		double delta_preferred_rack,
+		Cost_t core_transfer_cost,
+		Cost_t tor_transfer_cost,
+		uint32_t percent_block_tolerance,
+		uint64_t machines_per_rack)
+  : resource_map_(resource_map),
+    job_map_(job_map),
+    task_map_(task_map),
+    knowledge_base_(kb),
+		proportion_machine_preferred_(delta_preferred_machine),
+		proportion_rack_preferred_(delta_preferred_rack),
+		core_transfer_cost_(core_transfer_cost),
+		tor_transfer_cost_(tor_transfer_cost),
+		percent_block_tolerance_(percent_block_tolerance),
+		machines_per_rack_(machines_per_rack),
+		filesystem_(dfs),
+		runtime_distribution_(runtime_distribution),
+		block_distribution_(block_distribution)
+		{
+  // Shut up unused warnings for now
+  CHECK_NOTNULL(leaf_res_ids);
+  CHECK_NOTNULL(task_bindings);
+
+  // initialise to a single, empty rack
+  rack_to_machine_map_.assign(1, std::list<ResourceID_t>());
+}
+
+SimulatedQuincyCostModel::~SimulatedQuincyCostModel() {
+  for (auto elt : preferred_machine_map_) {
```

Comment:
"element", at a guess. But I've replaced it with "mapping" everywhere, which is clearer.